### PR TITLE
Add Johnson's algorithm for cycles in directed graph

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds a new function, simple_cycles(),  which is an
implementation of Johnson's algorithm [1] for finding all the simple
cycles in a directed graph. The implementation is based on the
non-recursive implementation from NetworkX. [2]

[1] https://doi.org/10.1137/0204007
[2] https://github.com/networkx/networkx/blob/networkx-2.8.4/networkx/algorithms/cycles.py#L98-L222

Closes: #622

TODO:

- [x] Add more tests

### CODE REVIEW

**Review by georgios-ts — COMMENTED** [review]
Thanks for doing this! I left some comments, most of them are minor. Besides that, this function should ideally return a custom return type that implements the iteration protocol and will lazily evaluate new simple cycles like we do in vf2 isomorphism https://github.com/Qiskit/retworkx/blob/main/src/isomorphism/vf2.rs#L978-L1048 since computing all cycles at once might consume more memory than it's really needed

File: src/connectivity/johnson_simple_cycles.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+) -> (StablePyGraph<Directed>, HashMap<NodeIndex, NodeIndex>) {
+    let node_set: HashSet<NodeIndex> = nodes.iter().copied().collect();
+    let mut node_map: HashMap<NodeIndex, NodeIndex> = HashMap::with_capacity(nodes.len());
+    let node_filter = |node: NodeIndex| -> bool { node_set.contains(&node) };
+    // Overallocates edges, but not a big deal as this is temporary for the lifetime of the
+    // subgraph
+    let mut out_graph = StablePyGraph::<Directed>::with_capacity(nodes.len(), graph.edge_count());
+    let filtered = NodeFiltered(&graph, node_filter);
+    for node in filtered.node_references() {
+        let new_node = out_graph.add_node(node.1.clone_ref(py));
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs:40]: That's a good point, I'll update this to use just a `StableGraph<(), ()>` I think to save some memory and avoid the need for any python when computing the cycles.

[the remark this answers is not in the document]
Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs]: Done in: https://github.com/Qiskit/rustworkx/pull/633/commits/38a124bdab88b3fcc808a6019ea881b9fdae91a4

File: src/connectivity/johnson_simple_cycles.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    }
+    (out_graph, node_map)
+}
+
+#[inline]
+fn set_pop(set: &mut HashSet<NodeIndex>) -> Option<NodeIndex> {
+    if set.is_empty() {
+        None
+    } else {
+        let set_node = set.iter().next().copied().unwrap();
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs]: Done in: https://github.com/Qiskit/rustworkx/pull/633/commits/790312091cb1d8bb08e406b709fd07e75befe384

File: src/connectivity/johnson_simple_cycles.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                                block.insert(neighbor, block_set);
+                            }
+                        };
+                    }
+                }
+                stack.pop();
+                path.pop();
+            }
+        }
+        let (h, node_map) = build_subgraph(
```
[the remark this answers is not in the document]
Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs]: Done in: https://github.com/Qiskit/rustworkx/pull/633/commits/38a124bdab88b3fcc808a6019ea881b9fdae91a4

**Comment by mtreinish** [comment]
I've also refactored the function to return a py iterator to compute a cycle on each iteration instead of doing a full computation and returning the whole output at once.

File: src/connectivity/johnson_simple_cycles.rs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            let cycle_node = self_cycles.pop().unwrap();
+            if self_cycles.is_empty() {
+                slf.self_cycles = None;
+            }
+            return Ok(IterNextOutput::Yield(NodeIndices {
+                nodes: vec![cycle_node.index()],
+            }));
+        }
+        // Restore previous state if it exists
+        let mut stack: Vec<(NodeIndex, IndexSet<NodeIndex>)> = slf.stack.drain(..).collect();
```
Comment (georgios-ts) [inline src/connectivity/johnson_simple_cycles.rs:209]: [`mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) might be more performant.
```suggestion
        let mut stack: Vec<(NodeIndex, IndexSet<NodeIndex>)> = mem::take(&mut slf.stack);
```

Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs:209]: oh, that's a good call. I'll update these to use `mem::take()`

Comment (mtreinish) [reply src/connectivity/johnson_simple_cycles.rs]: Done in: https://github.com/Qiskit/rustworkx/pull/633/commits/388111e2d1c5159364b9a4d82a603ec79749a9be

**Review by georgios-ts — APPROVED** [review]
Thanks for the updates. Looks good!

**Pull request merged**